src/future/join/tuple.rs

```diff
@@ -1,28 +1,37 @@
 use super::Join as JoinTrait;
-use crate::utils::MaybeDone;
+use crate::utils::PollState;
 
 use core::fmt::{self, Debug};
 use core::future::{Future, IntoFuture};
+use core::mem::MaybeUninit;
 use core::pin::Pin;
 use core::task::{Context, Poll};
 
-use pin_project::pin_project;
+use pin_project::{pin_project, pinned_drop};
 
-macro_rules! impl_merge_tuple {
+use paste::paste;
+
+macro_rules! impl_join_tuple {
     ($StructName:ident $($F:ident)*) => {
-        /// Waits for two similarly-typed futures to complete.
-        ///
-        /// This `struct` is created by the [`join`] method on the [`Join`] trait. See
-        /// its documentation for more.
-        ///
-        /// [`join`]: crate::future::Join::join
-        /// [`Join`]: crate::future::Join
-        #[pin_project]
-        #[must_use = "futures do nothing unless you `.await` or poll them"]
-        #[allow(non_snake_case)]
-        pub struct $StructName<$($F: Future),*> {
-            done: bool,
-            $(#[pin] $F: MaybeDone<$F>,)*
+        paste!{
+            /// Waits for two similarly-typed futures to complete.
+            ///
+            /// This `struct` is created by the [`join`] method on the [`Join`] trait. See
+            /// its documentation for more.
+            ///
+            /// [`join`]: crate::future::Join::join
+            /// [`Join`]: crate::future::Join
+            #[pin_project(PinnedDrop)]
+            #[must_use = "futures do nothing unless you `.await` or poll them"]
+            #[allow(non_snake_case)]
+            pub struct $StructName<$($F: Future),*> {
+                done: bool,
+                $(
+                    #[pin] $F: $F,
+                    [<$F _out>]: MaybeUninit<$F::Output>,
+                    [<$F _state>]: PollState,
+                )*
```

Comment:
We may want to create named fields here instead:

```suggestion
                futures: ($(#[pin]$F,)*),
                output: ($(MaybeUninit<$F::Output>,)*),
                state: ($(PollState,)*),
```